server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala

```diff
@@ -0,0 +1,136 @@
+package sttp.tapir.server.netty.cats.internal
+
+import cats.Applicative
+import cats.effect.kernel.{Async, Sync}
+import cats.effect.kernel.Resource.ExitCase
+import cats.effect.std.Dispatcher
+import cats.syntax.all._
+import fs2.interop.reactivestreams.{StreamSubscriber, StreamUnicastPublisher}
+import fs2.{Pipe, Stream}
+import io.netty.channel.ChannelPromise
+import io.netty.handler.codec.http.websocketx.{WebSocketFrame => NettyWebSocketFrame}
+import org.reactivestreams.{Processor, Publisher, Subscriber, Subscription}
+import org.slf4j.LoggerFactory
+import sttp.capabilities.fs2.Fs2Streams
+import sttp.tapir.model.WebSocketFrameDecodeFailure
+import sttp.tapir.server.netty.internal.WebSocketFrameConverters._
+import sttp.tapir.{DecodeResult, WebSocketBodyOutput}
+import sttp.ws.WebSocketFrame
+
+import scala.concurrent.ExecutionContext.Implicits
+import scala.concurrent.Promise
+import scala.util.{Failure, Success}
+
+/** A Reactive Streams Processor[NettyWebSocketFrame, NettyWebSocketFrame] built from a fs2.Pipe[F, REQ, RESP] passed from an WS endpoint.
+  */
+class WebSocketPipeProcessor[F[_]: Async, REQ, RESP](
+    pipe: Pipe[F, REQ, RESP],
+    dispatcher: Dispatcher[F],
+    o: WebSocketBodyOutput[Pipe[F, REQ, RESP], REQ, RESP, ?, Fs2Streams[F]],
+    wsCompletedPromise: ChannelPromise
+) extends Processor[NettyWebSocketFrame, NettyWebSocketFrame] {
+  @volatile private var subscriber: StreamSubscriber[F, NettyWebSocketFrame] = _
```

Comment:
It's set in `onSubscribe` which can be run from another thread. Netty internal reactive publisher and subscriber follow the same approach. They set other internal values without volatile, because they are updated/read only in onNext/onError etc, which are run always within Netty eventLoopExecutor, which guarantees the same thread (at least that's my understanding). Values set in `onSubscribe` don't have this guarantee.

Reply:
ok, maybe add this as a comment for your future self :)